Use libuv for tempdir function

musm opened the pull request:
Use libuv's implementation for `tempdir` for both linux and windows. This fixes the inconsistency between the windows and linux version, where the windows version does not currently include a trailing slash when calling `tempdir` compared to the linux implementation.

musm committed:
Use libuv API for tempdir function

Fixes the Windows tempdir function returning a path with the trailing slash.

musm commented:
Is this something we care about?

StefanKarpinski commented:
Yes, but too late for 1.2.

stevengj commented on base/file.jl:
```diff
@@ -424,7 +424,7 @@ function tempdir()
     lentemppath = ccall(:GetTempPathW, stdcall, UInt32, (UInt32, Ptr{UInt16}), length(temppath), temppath)
     windowserror("GetTempPath", lentemppath >= length(temppath) || lentemppath == 0)
-    resize!(temppath, lentemppath)
+    temppath[lentemppath] == 0x005c ?  resize!(temppath,lentemppath-1) :  resize!(temppath,lentemppath)
```
This would fail if `temppath` were the root directory, e.g. `C:\`, but I guess that shouldn't be possible.

musm replied:
Good point, this would occur if `TMP` env variable was set to `C:\\`

musm replied:
Latest commit checks that the folder is not a root directory and if it is a root directory does not strip the trailing slash

musm replied:
This fixes the case where `TMP` is set to `C:\\`  (kinda pathological, but in theory possible)

musm commented:
came up here https://discourse.julialang.org/t/tempdir-differences-in-windows-and-linux/24930

musm commented on base/file.jl:
```diff
@@ -424,7 +424,8 @@ function tempdir()
     windowserror("GetTempPath", lentemppath >= length(temppath) || lentemppath == 0)
-    resize!(temppath, lentemppath)
+    isroot = ccall((:PathIsRootW,:Shlwapi), stdcall, Int32, (Ptr{UInt16},), temppath)
+    isroot != 1 && temppath[lentemppath] == 0x005c ?  resize!(temppath,lentemppath-1) :  resize!(temppath,lentemppath)
```
I can also write this is as
```julia
    isroot = Bool(ccall((:PathIsRootW,:Shlwapi), stdcall, Cint, (Ptr{UInt16},), temppath))
    !isroot && temppath[lentemppath] == 0x005c ?  resize!(temppath,lentemppath-1) :  resize!(temppath,lentemppath)
```

stevengj replied:
Or
```jl
isroot = Bool(ccall((:PathIsRootW,:Shlwapi), stdcall, Cint, (Ptr{UInt16},), temppath))
resize!(temppath, !isroot && temppath[lentemppath] == 0x005c ? lentemppath-1 : lentemppath)
```

vtjnash replied:
Or
```julia
isroot = Bool(ccall((:PathIsRootW, :Shlwapi), stdcall, Cint, (Ptr{UInt16},), temppath))
(!isroot && temppath[lentemppath] == 0x005c) && (lentemppath -= 1)
resize!(temppath, lentemppath)
```

vtjnash commented on base/file.jl:
```diff
@@ -424,7 +424,8 @@ function tempdir()
     lentemppath = ccall(:GetTempPathW, stdcall, UInt32, (UInt32, Ptr{UInt16}), length(temppath), temppath)
     windowserror("GetTempPath", lentemppath >= length(temppath) || lentemppath == 0)
-    resize!(temppath, lentemppath)
+    isroot = ccall((:PathIsRootW,:Shlwapi), stdcall, Int32, (Ptr{UInt16},), temppath)
```
`Shlwapi` is a very large dependency that we don't appear to use anywhere else. It would be nice avoid instantiating it just for this. There should already be a regex for this in path.jl for splitdrive.

musm replied:
Sure, is there any overhead to instatitating it or is the preference to keep as much of the code in Julia? I don't see any real overhead to calling the library since it's loaded by default.

musm replied:
Updated code with a julia version of `isroot`. Let me know if this is preffered

vtjnash commented on test/file.jl:
```diff
@@ -246,6 +246,8 @@ end
 my_tempdir = tempdir()
 @test isdir(my_tempdir) == true
+@test  my_tempdir[end] != "/"
+@test  my_tempdir[end] != "\\"
```
excess space?

musm commented on base/file.jl:
```diff
@@ -420,12 +420,19 @@ const temp_prefix = "jl_"
 
 if Sys.iswindows()
 
+function _isroot(path::String)
```
Should I move this to `path.jl` ?

stevengj commented on base/file.jl:
```diff
@@ -420,13 +420,19 @@ const temp_prefix = "jl_"
+    resize!(temppath,lentemppath)
+    path = transcode(String, temppath)
+    # ignore trailing `\` except for drive-root
+    return _isroot(path) ?  path : path[end] == '\\' ? path[1:end-1] : path
```
Isn't the common case for `temppath` to end with `\`, which means that this makes an extra copy of the string in the common case?

musm replied:
Yea that is the common case. I'm not sure how'd go about avoiding the extra copy when using the julia  `_isroot` implementation.

stevengj replied:
We could use a modified isroot regex that is missing the trailing slash, and test it on the string after we strip the slash—if it matches, add the slash back.

musm commented on base/file.jl:
```diff
@@ -420,12 +420,20 @@ const temp_prefix = "jl_"
 
 if Sys.iswindows()
 
+function ispathroot(path::String)
```
Should I move this to `path.jl`?

stevengj commented on base/file.jl:
```diff
@@ -420,19 +420,20 @@ const temp_prefix = "jl_"
     lentemppath = ccall(:GetTempPathW, stdcall, UInt32, (UInt32, Ptr{UInt16}), length(temppath), temppath)
     windowserror("GetTempPath", lentemppath >= length(temppath) || lentemppath == 0)
-    resize!(temppath,lentemppath)
+    resize!(temppath,lentemppath-1) # ignore trailing `\`
```
Shouldn't it check that there is a trailing `\` first?  e.g.
```jl
resize!(temppath,lentemppath - (temppath[lentemppath]==0x005c))
```

musm replied:
According to their docs https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-gettemppathw
the returned string ends with a backslash.

stevengj replied:
I like how their docs say that it ends with a backslash and immediately contradicts itself with the example: ``` The returned string ends with a backslash, for example, "C:\TEMP". ```

musm replied:
I know, someone at MSFT needs to fix that....    (I tested several combinations and the function always adds backslash...)

musm replied:
Apparently already reported https://github.com/MicrosoftDocs/feedback/issues/898

stevengj commented on base/file.jl:
```diff
@@ -420,19 +420,20 @@ const temp_prefix = "jl_"
+function ispathroot(path::String)
     m = match(r"^([^\\]+:|\\\\[^\\]+\\[^\\]+|\\\\\?\\UNC\\[^\\]+\\[^\\]+|\\\\\?\\[^\\]+:|)(.*)$", path)
-    m.captures[2] == "\\" ? true : false
+    (m.captures[2] == "\\" || isempty(m.captures[2])) ? true : false
```
If `m.captures[2] == "\\"` then we don't need to add another `\`, right?    So we actually just want `isempty(m.captures[2])` for this case.    You also don't need the `... ? true : false`, since you can just return `...`.

Maybe it's better to just inline
```jl
# check whether path is the drive root or UNC path, missing the trailing backslash:
isroot = isempty(match(r"^([^\\]+:|\\\\[^\\]+\\[^\\]+|\\\\\?\\UNC\\[^\\]+\\[^\\]+|\\\\\?\\[^\\]+:|)(.*)$", path).captures[2])
```

musm replied:
if `path = "C:"` then   `m.captures[2] = ""`   if `path = "C:\\"` then  `m.captures[2]== "\\"` . Path paths are roots.      Because `GetTempPathW` always adds a backslash `isempty(m.captures[2])`  is not needed, however I added it since the function tests if the path is rooted.

musm replied:
I'm not sure how useful this function is in general, but it could be made to check is the path is a root for   non -windows OS,   by checking if the path starts with `/` ?

musm commented:
@stevengj I addressed your feedback, let me know if these changes now have your approval. Thanks for the review.


I also think GetTempPathW could probably just be implemented in Julia, without too much difficulty. Probably best for another PR, however.

> 
> The GetTempPath function checks for the existence of environment variables in the following order and uses the first path found:
> 
> The path specified by the TMP environment variable.
> The path specified by the TEMP environment variable.
> The path specified by the USERPROFILE environment variable.
> The Windows directory.
> Note that the function does not verify that the path exists, nor does it test to see if the current process has any kind of access rights to the path. The GetTempPath function returns the properly formatted string that specifies the fully qualified path based on the environment variable search order as previously specified. The application should verify the existence of the path and adequate access rights to the path prior to any use for file I/O operations.
> Symbolic link behavior—If the path points to a symbolic link, the temp path name maintains any symbolic links.
>

vtjnash commented:
If we want to go that route, we could switch to just using `uv_os_tmpdir` everywhere (it already handle this case).

musm commented:
Any reason not to go the libuv route? Seems simpler, then custom casing Windows.

musm commented:
I was testing `uv_os_tmpdir`

```julia
function _tempdir()
    path_max = 1024
    buf = zeros(UInt8, path_max)
    sz = Ref{Csize_t}(path_max)
    rc = ccall(:uv_os_tmpdir, Cint, (Ptr{UInt8}, Ptr{Csize_t}), buf, sz)
    if rc == 0
        @show buf, sz[]
        resize!(buf, sz[])
        return String(buf)
    else
        error("unable to retrieve temporary directory directory")
    end
end
  
```

it seems to work fine EXCEPT when `ENV["TMP"]= ""` ,  in which case garbage is spit out. `GetTempPathW ` would spit out the correct value.  I think there might be a bug in  `uv_os_tmpdir` ?

stevengj commented on base/file.jl:
```diff
@@ -416,18 +416,29 @@ function touch(path::AbstractString)
-    resize!(temppath, lentemppath)
-    return transcode(String, temppath)
+    path_max = 1024
+    buf = Vector{UInt8}(undef, path_max)
```
StringVector?

musm replied:
sure

stevengj commented on base/file.jl:
```diff
@@ -416,18 +416,29 @@ function touch(path::AbstractString)
+    buf = Vector{UInt8}(undef, path_max)
+    sz = Ref{Csize_t}(path_max)
+    rc = ccall(:uv_os_tmpdir, Cint, (Ptr{UInt8}, Ptr{Csize_t}), buf, sz)
+    @assert rc == 0
```
Assertion or systemerror?

musm replied:
I'm not 100% sure which is best.  There are two scenarios were `rc!=0`: Invalid inputs to `uv_os_tmpdir`, for which `@assert` is appropriate since this should never happen; and  possible errors  related to buffer having e.g. too small a length, which shouldn't happen, but might, for which throwing an error is probably best?

From the docs

```
 On success size is set to the string length of buffer (which does not include the terminating null). On UV_ENOBUFS failure size is set to the required length for buffer, including the null byte.
```

musm commented:
According to https://github.com/JuliaLang/julia/commit/84e68f3a7ef7d888af8a580dc52f27ff9209f4ef by @Keno  it seems like we should be using RefValue instead of Ref here?  (there are to instances where Ref is used)

stevengj commented on base/file.jl:
```diff
@@ -416,18 +416,29 @@ function touch(path::AbstractString)
+    buf = Base.StringVector(path_max)
+    sz = Ref{Csize_t}(path_max + 1)
+    rc = ccall(:uv_os_tmpdir, Cint, (Ptr{UInt8}, Ptr{Csize_t}), buf, sz)
+    rc != 0 && error("unable to retrieve temporary directory")
```
Use e.g. `uv_error("tempdir", rc)`, which checks the error code and gives an appropriate message.

musm commented:
Ok I updated the code and this is now ready for a final review. Last commit handles  the UV_ENOBUFS case with tests. I think this is now good to merge.

stevengj commented on base/file.jl:
```diff
@@ -416,18 +416,36 @@ function touch(path::AbstractString)
-    resize!(temppath, lentemppath)
-    return transcode(String, temppath)
+    path_max = 1024
+    buf = Base.StringVector(path_max)
```
needs to be length `path_max+1` since libuv copies the trailing NUL char: https://github.com/JuliaLang/libuv/blob/26dbe5672c33fc885462c509fe2a9b36f35866fd/src/unix/core.c#L1105-L1106

stevengj replied:
(either that or set `sz` to `path_max` … it's okay if its too small since we reallocate in the unlikely event of such a long path.)

musm replied:
Thanks.
I removed the `sz` initialization, since it's not needed.

stevengj commented on base/file.jl:
```diff
@@ -416,18 +416,36 @@ function touch(path::AbstractString)
+            resize!(buf, sz[])
+            return String(buf)
+        elseif rc == Base.UV_ENOBUFS
+            resize!(buf, sz[] - 1)
```
No `- 1`, similar to above.

musm commented:
unrelated error in testset Distributed on macos

musm commented on test/file.jl:
```diff
@@ -244,8 +244,22 @@ close(s)
+
+    var =  Sys.iswindows() ? "TMP" : "TMPDIR"
+    PATH_PREFIX = Sys.iswindows() ? "C:\\" : "/tmp/"
+    MAX_PATH = (Sys.iswindows() ? 251 : 1024) - length(PATH_PREFIX) - 1 # null-termination character
```
For reference I'm pretty sure there's a bug in the libuv function that hits an edge case that shouldn't happen, which is why the test here is for 251 chars and does not test longer paths.

hitting this branch I believe https://github.com/libuv/libuv/blob/c4e9657d59723a6273ce68bda935680e1b3596c5/src/win/util.c#L1186
Looks like the issue is that the code is using MAXPATH+1 instead of MAXPATH

musm commented:
Before merging let me push an update to account for the latest changes on master

musm commented:
Test failures are flukes (distributed fluke and a timeout fluke during distributed). Is this  ok to merge ?

musm commented:
good to merge?

musm commented:
Test failures are unrelated, due to downloads

musm committed:
Simplify comments and behavior

musm commented:
@vtjnash can you sign off if this looks good to you? 

FYI the `@test_broken` is fixed by my PR to libuv

vtjnash commented on base/file.jl:
```diff
@@ -481,9 +497,6 @@ function tempname()
     return s
 end
```
Change to use `tempdir()` for `d`?

musm replied:
Will make these changes in a separate PR.

vtjnash replied:
sgtm. feel free to merge.

vtjnash commented on base/file.jl:
```diff
@@ -496,13 +509,6 @@ end
-"""
-tempdir()
-
 """
```
In the `tempname` doc string below, it might be clearer to add, "This can lead to security holes." and recommend opening the file with `O_EXCL`, if this is a concern (in addition to the suggestion to use `mktemp`).

Aside: the windows implementation of `mktemp` appears to be wrong because it fails to heed this warning.

musm commented:
> sgtm. feel free to merge.


Someone with merge powers want to take care of this 😉
